(PUP-9024) tagged: downcase arguments before checking

mateusz-gozdek-sociomantic opened the pull request:
As tags are always stored downcased, we should also downcase
passed arguments, otherwise function will never return true.

puppetlabs-jenkins commented:
Can one of the admins verify this patch?

puppetcla commented:
CLA signed by all contributors.

joshcooper commented:
Thanks for your contribution @mateusz-gozdek-sociomantic. I think we can simplify this even more, because the `Puppet::Util::Tagging` module is already mixed into the resource and catalog, and it handles the downcasing and testing an array of values. So you should be able to replace most of the function with just:

```
compiler.catalog.tagged?(vals) || resource.tagged?(vals)
```

Also it would be good to add a unit test to make sure the `tagged` function is case-insensitive.

joshcooper commented:
We should also target 5.5.x as the base branch. You should be able to edit this PR to change the branch instead of opening a new PR.

mateusz-gozdek-sociomantic commented:
Thanks for the feedback @joshcooper! I changed base branch, rebased and applied your suggestion, however it seems, that your solution is not working, but I'll try to figure that out.

I added test for this case, but I have no experience in it, so I'd appreciate some help/suggestions. What I have I copied from `tag_spec.rb`.

mateusz-gozdek-sociomantic commented:
Another thing is, that `tagged` function is ANDing the parameters:
> The tags are ANDed, so that all of the specified tags must be included for
  the function to return true.

And `raw_tagged?`: 
> Answers if this resource is tagged with at least one of the tags given in downcased string form.

So I guess this should also be added to tests.

mateusz-gozdek-sociomantic commented:
I updated the logic, which works as expected, but I need some help with tests, as I can't figure out why they are failing :disappointed:

melissa commented on spec/unit/parser/functions/tagged_spec.rb:
```diff
@@ -22,4 +22,20 @@
+
+  it 'should check if all specified tags are included' do
+    resource = Puppet::Parser::Resource.new(:file, "/file", :scope => @scope)
+    @scope.expects(:resource).returns resource
```
I believe these tests are both failing because we want `stubs` here rather than `expects`. Where `expects` expects the method to be called exactly once, or a specified set of times, `stubs` will accept the method any number of times. Once I made that change, these tests passed for me.

mateusz-gozdek-sociomantic replied:
Thanks for suggestion @melissa, tests are indeed passing now.

joshcooper commented:
jenkins please test this

joshcooper commented:
```
$ bundle exec puppet apply manifest.pp
Notice: Compiled catalog for localhost in environment production in 0.03 seconds
Notice: true
Notice: /Stage[main]/Main/Notify[true]/message: defined 'message' as 'true'
Notice: Applied catalog in 0.01 seconds
```
